pgrx/src/callconv.rs

```diff
@@ -134,3 +314,325 @@ fn finish_srf_init<T>(
+{
+    type CallRet = T::CallRet;
+
+    unsafe fn prepare_call(fcinfo: pg_sys::FunctionCallInfo) -> CallCx {
+        T::prepare_call(fcinfo)
+    }
+
+    fn label_ret(self) -> Ret<Self> {
+        let value = pg_sys::panic::ErrorReportable::unwrap_or_report(self);
+        match T::label_ret(value) {
+            Ret::Zero => Ret::Zero,
+            Ret::Once(value) => Ret::Once(value),
+            Ret::Many(iter, value) => Ret::Many(Ok(iter), value),
+        }
+    }
+
+    unsafe fn box_return(fcinfo: pg_sys::FunctionCallInfo, ret: Ret<Self>) -> pg_sys::Datum {
+        let ret = match ret {
+            Ret::Zero => Ret::Zero,
+            Ret::Once(value) => Ret::Once(value),
+            Ret::Many(iter, value) => {
+                let iter = pg_sys::panic::ErrorReportable::unwrap_or_report(iter);
+                Ret::Many(iter, value)
+            }
+        };
+
+        T::box_return(fcinfo, ret)
+    }
+
+    unsafe fn into_context(self, fcinfo: pg_sys::FunctionCallInfo) {
+        match self {
+            Err(_) => (),
+            Ok(value) => value.into_context(fcinfo),
+        }
+    }
+
+    unsafe fn ret_from_context(fcinfo: pg_sys::FunctionCallInfo) -> Ret<Self> {
+        match T::ret_from_context(fcinfo) {
+            Ret::Many(iter, value) => Ret::Many(Ok(iter), value),
+            Ret::Once(value) => Ret::Once(value),
+            Ret::Zero => Ret::Zero,
+        }
+    }
+
+    unsafe fn finish_call(fcinfo: pg_sys::FunctionCallInfo) {
+        T::finish_call(fcinfo)
+    }
+}
+
+macro_rules! impl_boxret_for_primitives {
+    ($($scalar:ty),*) => {
+        $(
+        unsafe impl ReturnShipping for $scalar {
+            type CallRet = Self;
+            fn label_ret(self) -> Ret<Self> {
+                Ret::Once(self)
+            }
+
+            unsafe fn box_return(fcinfo: pg_sys::FunctionCallInfo, ret: Ret<Self>) -> pg_sys::Datum {
+                match ret {
+                    Ret::Zero => unsafe { pg_return_null(fcinfo) },
+                    Ret::Once(value) => $crate::pg_sys::Datum::from(value),
+                    Ret::Many(_, _) => unreachable!()
+                }
+            }
+        }
+        )*
+    }
+}
+
+impl_boxret_for_primitives! {
+    i8, i16, i32, i64, bool
+}
+
+unsafe impl ReturnShipping for () {
+    type CallRet = Self;
+    fn label_ret(self) -> Ret<Self> {
+        Ret::Zero
+    }
+
+    unsafe fn box_return(fcinfo: pg_sys::FunctionCallInfo, ret: Ret<Self>) -> pg_sys::Datum {
+        pg_sys::Datum::from(0)
+    }
+}
+
+unsafe impl ReturnShipping for f32 {
+    type CallRet = Self;
+    fn label_ret(self) -> Ret<Self> {
+        Ret::Once(self)
+    }
+
+    unsafe fn box_return(fcinfo: pg_sys::FunctionCallInfo, ret: Ret<Self>) -> pg_sys::Datum {
+        match ret {
+            Ret::Zero => unsafe { pg_return_null(fcinfo) },
+            Ret::Once(value) => pg_sys::Datum::from(value.to_bits()),
+            Ret::Many(_, _) => unreachable!(),
+        }
+    }
```

Comment:
This is what I meant about the awkwardness of forcing the scalar type cases into a framework that can support the iterator and composite type cases.

Reply:
I resolved this by using a blanket impl.